File: lib/to_mdx/parsers/open_document/text/list.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Text
      class List < Base
        attr_accessor :style_name

        def initialize(element = nil, context = {})
          super(element, context)
          @style_name = context[:styles].use(element&.attributes&.[]("text:style-name"))
        end

        def to_mdx
          style_class = @style_name.empty? ? "" : " className=\"#{@style_name}\""
          <<~MDX
            <ul#{style_class}>
              #{contentful_children.map(&:to_mdx).join}
            </ul>
          MDX
        end

        def to_s
          contentful_children.map(&:to_s).join("\n")
        end

        register_for "text:list"
      end
    end
  end
end

File: src/parsers/open-document/text/list.ts
import dedent from "dedent-js";
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class List extends BaseElement {
  private styleName: Style | undefined = undefined;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.styleName = context.styles.use(this.attr("text:style-name"));
  }

  toMdx(): string {
    const styleClass = this.styleName?.empty() ? "" : ` className="${this.styleName}"`;
    const content = this.contentfulChildren()
      .map((child) => child.toMdx())
      .join("");

    return dedent`
      <ul${styleClass}>
        ${content}
      </ul>
    `;
  }

  toString(): string {
    return this.contentfulChildren()
      .map((child) => child.toString())
      .join("\n");
  }

  static {
    BaseElement.registerFor.call(this, "text:list");
  }
}
